### ISSUE TITLE
[DI] Mixing constructor injection and setter injection breaks php dumper

### ISSUE THREAD

**Issue by discordier** [symfony/symfony#28824]
**Symfony version(s) affected**: 4.1 (since 4.1.5 - previous not affected)

**Description**  
When having a "complex" service graph consisting a mixture of constructor injection and setter injection, the dumped function in the container will attempt to use a variable before initializing it.

The container will error out with something like:
```
Notice: Undefined variable: b
```

**How to reproduce**  
I tried to reduce the amount of involved services to the absolute minimum, this is how far I got it narrowed down.
The following unit test breaks the container dumper on my machine.
Reducing it more, like removing any of the `->addMethodCall()` declarations or any of the arguments of `multiuse1`, will not exhibit the issue anymore.
```diff
diff --git a/src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php b/src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php
index 8cb11568f2..2767fd9847 100644
--- a/src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php
+++ b/src/Symfony/Component/DependencyInjection/Tests/Dumper/PhpDumperTest.php
@@ -1071,6 +1071,49 @@ class PhpDumperTest extends TestCase
         $container = new \Symfony_DI_PhpDumper_Errored_Definition();
         $container->get('runtime_error');
     }
+
+    public function testCircularWithAddCalls()
+    {
+        $container = new ContainerBuilder();
+
+        $container
+            ->register('root', 'stdClass')
+            ->setArguments([new Reference('level2'), new Reference('multiuse1')])
+            ->setPublic(true);
+
+        $container
+            ->register('level2', FooForCircularWithAddCalls::class)
+            ->addMethodCall('call', [new Reference('level3')]);
+
+        $container->register('multiuse1', 'stdClass');
+
+        $container
+            ->register('level3', 'stdClass')
+            ->addArgument(new Reference('level4'));
+
+        $container
+            ->register('level4', 'stdClass')
+            ->setArguments([new Reference('multiuse1'), new Reference('level5')]);
+
+        $container
+            ->register('level5', 'stdClass')
+            ->addArgument(new Reference('level6'));
+
+        $container
+            ->register('level6', FooForCircularWithAddCalls::class)
+            ->addMethodCall('call', [new Reference('level5')]);
+
+        $container->compile();
+        $dumper = new PhpDumper($container);
+
+        $content = $dumper->dump(array('class' => 'Symfony_DI_PhpDumper_Test_Circular_With_Add_Calls'));
+
+        eval('?>'.$content);
+
+        $container = new \Symfony_DI_PhpDumper_Test_Circular_With_Add_Calls();
+
+        $this->assertInstanceOf(\stdClass::class, $container->get('root'));
+    }
 }
 
 class Rot13EnvVarProcessor implements EnvVarProcessorInterface
@@ -1096,3 +1139,10 @@ class FooForDeepGraph
         $this->bClone = clone $b;
     }
 }
+
+class FooForCircularWithAddCalls
+{
+    public function call(\stdClass $argument)
+    {
+    }
+}
```

**Comment by nicolas-grekas** [comment]
This should be already fixed in `4.1@dev`, can you please confirm?

**Comment by discordier** [comment]
I get test failure on 61cf143727 
```
$ git describe 
v4.1.6-52-g61cf143727
```

**Comment by nicolas-grekas** [comment]
I meant 4.1@dev for the symfony/dependency-injection component

**Comment by discordier** [comment]
Still the same problem in https://github.com/symfony/dependency-injection/commit/faaa406b967161d01b56f08645dec46699eaad93
```
git describe 
v4.1.6-2-gfaaa406b
```

**Comment by discordier** [comment]
bisect lead me to https://github.com/symfony/dependency-injection/commit/ef44ed4b69c8deacb7c5e171bd87a950c6777c1c which is the first bad commit.

**Comment by pkruithof** [comment]
I have this issue too, though without any `addMethodCall` declarations. Here is the (truncated and redacted) compiled file:

```php
<?php

use Symfony\Component\DependencyInjection\Argument\RewindableGenerator;
use Symfony\Component\DependencyInjection\Exception\RuntimeException;

// This file has been auto-generated by the Symfony Dependency Injection Component for internal use.
// Returns the private 'feed.command_bus.prooph' shared autowired service.

$a = ($this->privates['Some\Class'] ?? $this->load('getSomeService.php'));

if (isset($this->privates['feed.command_bus.prooph'])) {
    return $this->privates['feed.command_bus.prooph'];
}

$c = ($this->privates['Prooph\EventStore\ActionEventEmitterEventStore'] ?? $this->load('getActionEventEmitterEventStoreService.php'));

if (isset($this->privates['feed.command_bus.prooph'])) {
    return $this->privates['feed.command_bus.prooph'];
}
$b = \Some\Class::create($c, $this->getEnv('foo'));
$e = ($this->privates['Some\Other\Class'] ?? $this->load('getSomeOtherService.php'));

if (isset($this->privates['feed.command_bus.prooph'])) {
    return $this->privates['feed.command_bus.prooph'];
}
$f = ($this->privates['Another\Class'] ?? $this->load('getAnotherService.php'));

if (isset($this->privates['feed.command_bus.prooph'])) {
    return $this->privates['feed.command_bus.prooph'];
}

$this->privates['feed.command_bus.prooph'] = $instance = \CommandBus\Factory::create(array(
    'Some\\CommandA' => new CommandHandler($a), 
    'Some\\CommandB' => new CommandHandler($b),
    'Some\\CommandC' => new CommandHandler($d),
    'Some\\CommandD' => new CommandHandler($e),
    'Some\\CommandE' => new CommandHandler($f), 
    'Some\\CommandF' => new CommandHandler($g), 
);

$d = \Some\Other\Class::create($c, $this->getEnv('bar'));

$g = \Another\Class::create($c, $this->getEnv('baz'));

return $instance;
```

A couple of things:

* A check if the service is already in privates keeps repeating (`if (isset(service)) { return service }`)
* The dependencies are out of order (`$c` before `$b`), but maybe that's not a bad thing, as long as the service itself is created last. Which is problem #3
* After the service creation, dependencies are created.

**Comment by nicolas-grekas** [comment]
Can you please share the dumped container as xml with me please, eg using slack?

**Comment by discordier** [comment]
@nicolas-grekas: Do you need anything else from me or is my unit test above enough?

**Comment by nicolas-grekas** [comment]
I need to check your unit test first, thanks.

**Comment by pkruithof** [comment]
Any news on this? A second project of mine now also has a broken container 😢

**Comment by nicolas-grekas** [comment]
I'm having a look early next week.

**Comment by nicolas-grekas** [comment]
Should be fixed in #29103

**Comment by nicolas-grekas** [comment]
(now merged up to master, should be easier to test)
